File: python/house_robber.py
# no. 198
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = nums[1]
        for i in range(2, len(nums)):
            if i == 2:
                dp[i] = dp[0] + nums[i]
            else:
                dp[i] = max(dp[i - 2], dp[i - 3]) + nums[i]
        return max(dp[-1], dp[-2])

File: C++/house_robber.cpp
// no. 198
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
  int rob(vector<int> &nums) {
    if (nums.size() == 1) {
      return nums[0];
    }
    vector<int> dp;
    dp.push_back(nums[0]);
    dp.push_back(nums[1]);
    for (int i = 2; i < nums.size(); i++) {
      if (i == 2) {
        dp.push_back(dp[0] + nums[i]);
      } else {
        dp.push_back(max(dp[i - 2], dp[i - 3]) + nums[i]);
      }
    }
    return max(dp[nums.size() - 2], dp[nums.size() - 1]);
  }
};
